substrate/client/network/sync/src/strategy/chain_sync.rs

```diff
@@ -2048,16 +2052,23 @@ fn peer_block_request<B: BlockT>(
 	max_parallel_downloads: u32,
 	max_blocks_per_request: u32,
 	finalized: NumberFor<B>,
-	best_num: NumberFor<B>,
+	best_queued_num: NumberFor<B>,
+	best_number: NumberFor<B>,
+	best_hash: B::Hash,
 ) -> Option<(Range<NumberFor<B>>, BlockRequest<B>)> {
-	if best_num >= peer.best_number {
+	// Nothing to download from the peer via normal block requests.
+	if best_number == peer.best_number && best_hash == peer.best_hash {
```

Comment:
What if `peer.best_number > best_number`? The fix won't work and the duplicate block request will be issued?

Reply:
Good point. The fix is incomplete in this case, the block requests would include both the duplicate and new blocks, so the gap sync will still be interrupted unexpectedly. I'll need some more time to think about it.